quota/etcd/quotaapi/name.go

```diff
@@ -0,0 +1,94 @@
+// Copyright 2017 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package quotaapi
+
+import (
+	"fmt"
+	"regexp"
+	"strconv"
+	"strings"
+)
+
+const (
+	collectionTrees = "trees"
+	pathWildcard    = "-"
+)
+
+var (
+	// globalRE is a regex for quotas/global/ name filters.
+	// Wildcards ("-") are not allowed to replace "global". They could be allowed, but it's a query
+	// that doesn't make much sense, as if this format is used it has to be a global quota query.
+	globalRE *regexp.Regexp
+
+	// treeUsersRE is a broader regex for all trees/ and users/ name filters.
+	// It doesn't guard against the following specific situations:
+	// 1. quotas/trees/<id>/<kind>/config where id is not an int64
+	// 2. quotas/-/<id>/<kind>/config where id != "-", which is a possibly ambiguous query. It could
+	//    be allowed, but since the ID being queried is already known, so is the collection, so it
+	//    doesn't make much sense.
```

Comment:
Tree ID has to be an int64. User ID can be whatever you can fit in an URL, minus "/" (and "-", now that I think of it). For the latter, I honestly don't think we should open the can of worms and attempt to validate what can go in a URL and what can't.

Think of this as a "best-effort" validation: if a name looks wrong and we can tell it, we won't do the search. If a name has esoteric chars but passes the regex, we'll do the search (and possibly come back with an empty response). If someone creates config names with emoticons or whatever else they fancy, and can get it through gRPC Proxy, well, who am I to stop them?  ;)

The intent is to alert people querying clearly bad names. We could forgo all validation, attempt the matches and come back empty handed, but that seemed... unhelpful.

I considered multiple regexes at first. I ended up with 4 of them: one for globals, one for trees, one for users and one for "trees+users" (ie, `quotas/-/-/...`). It seemed a bit too many, plus `\d+` is not equivalent to `ParseInt(_, 10, 64)`, so in the end I opted for less regexes and a bit of handcrafted logic.